src/Orchestration/Adapter/DockerCLI.php

```diff
@@ -4,30 +4,145 @@
 
 use Utopia\CLI\Console;
 use Utopia\Orchestration\Adapter;
+use Utopia\Orchestration\Container;
+use Utopia\Orchestration\Exceptions\DockerCLIException;
+use Utopia\Orchestration\Exceptions\TimeoutException;
 
 class DockerCLI extends Adapter
 {
-    public function __construct($username, $password)
+    /**
+     * Constructor
+     * 
+     * @param string $username
+     * @param string $password
+     * @return void
+     */
+    public function __construct(string $username = null, string $password = null)
     {
-        if($username) {
-            Console::execute('docker login --username '.$username.' --password-stdin', $password, '', '');
+        if($username && $password) {
+            $stdout = '';
+            $stderr = '';
+
+            if (!Console::execute('docker login --username '.$username.' --password-stdin', $password, $stdout, $stderr)) {
+                throw new DockerCLIException("Docker Error: {$stderr}");
+            };
         }
     }
 
+
+    /**
+     * Pull Image
+     * 
+     * @param string $image
+     * 
+     * @return bool
+     */
     public function pull(string $image): bool
     {
-        return (Console::execute('docker pull '.$image, '', '', '') === 0);
+        $stdout = '';
+        $stderr = '';
+
+        $result = Console::execute('docker pull '.$image, '', $stdout, $stderr);
+
+        if ($result !== 0) {
+            throw new DockerCLIException("Docker Error: {$stderr}");
+        }
+
+        return !$result;
     }
 
-    public function list(): bool
+    /**
+     * List Containers
+     * @param array<string, string> $filters
+     *
+     * @return Container[]
+     */
+    public function list(array $filters = []): array
     {
-        return (Console::execute('docker ps --all --format "name={{.Names}}&status={{.Status}}&labels={{.Labels}}" --filter label='.$this->namespace.'-type=runtime', '', '', '') === 0);
+        $stdout = '';
+        $stderr = '';
+
+        $filterString = '';
+        array_walk($filters,
```

Comment:
Instead of `array_walk`, we can use `array_reduce` here to prevent having to `use (&$filterString)`